### PR TITLE
Modulerize records package

### PR DESCRIPTION
**Pull request by sttts**
- split `InsertState` as much as possible
- move all state.json logic into subpackage `state`

This is the preparation for the DiscoveryInfo support.

### CODE REVIEW

File: records/generator.go [hunk 1]
```diff
@@ -244,106 +183,105 @@ func hostToIP4(hostname string) (string, bool) {
...
-			address = labels.DomainFrag(address, labels.Sep, spec)
-		}
-		rg.SlaveIPs[slave.ID] = address
-	}
+	rg.frameworkRecords(sj, domain, spec)
+	rg.slaveRecords(sj, domain, spec)
+	rg.listenerRecord(listener, ns)
+	rg.masterRecord(domain, masters, sj.Leader)
 
 	// complete crap - refactor me
```
Comment (tsenart) [inline records/generator.go]: Why don't we have an `rg.taskRecords(...)` ?

File: records/generator.go [hunk 2]
```diff
@@ -244,106 +183,105 @@ func hostToIP4(hostname string) (string, bool) {
...
-				for _, port := range ports {
-					srvhost := trec + ":" + port
-					tcp := "_" + tname + "._tcp." + tail
-					udp := "_" + tname + "._udp." + tail
-					rg.insertRR(tcp, srvhost, "SRV")
-					rg.insertRR(udp, srvhost, "SRV")
-				}
-			}
+			// Add RFC 2782 SRV records
+			task.ForEachPort(func(port string) {
```
Comment (tsenart) [inline records/generator.go]: This sort of closure iteration isn't very idiomatic in Go. What prevents us from having a method which simply returns a slice of ports and ranging over that instead?

File: records/generator.go [hunk 3]
```diff
@@ -244,106 +183,105 @@ func hostToIP4(hostname string) (string, bool) {
...
 			if !ok || (task.State != "TASK_RUNNING") {
 				continue
 			}
 
-			tname := spec(task.Name)
-			sid := slaveIDTail(task.SlaveID)
-			tag := hashString(task.ID)
-			tail := fname + "." + domain + "."
+			// context used to build domain names
+			context := struct {
```
Comment (tsenart) [inline records/generator.go]: s/context/ctx/`? It'd make lines shorter without sacrificing any information value.

Comment (tsenart) [inline records/generator.go]: Also, more compact:
```go
ctx := struct{ TaskName, TaskID, SlaveID string }{
	spec(task.Name), hashString(task.ID), slaveIDTail(task.SlaveID),
}
```

File: records/generator.go [hunk 4]
```diff
@@ -244,106 +183,105 @@ func hostToIP4(hostname string) (string, bool) {
...
+			}{
+				SlaveID:  slaveIDTail(task.SlaveID),
+				TaskID:   hashString(task.ID),
+				TaskName: spec(task.Name),
+			}
 
-			// A records for task and task-sid
-			arec := tname + "." + tail
+			// insert canonical A records
+			trec := context.TaskName + "-" + context.TaskID + "-" + context.SlaveID + "." + tail
```
Comment (tsenart) [inline records/generator.go]: Eventually I'd like to have these name constructions extracted into well tested, single purpose functions.

File: records/generator_test.go [hunk 5]
```diff
@@ -261,19 +212,19 @@ func TestInsertState(t *testing.T) {
 }
 
 func TestPID_UnmarshalJSON(t *testing.T) {
```
Comment (tsenart) [inline records/generator_test.go]: This test should move to the `state` package.

File: records/state/utils.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
+package state
```
Comment (tsenart) [inline records/state/utils.go]: IMO `utils`, `misc` and such packages are a smell. They're the trash bin of programming taxonomy. :-)

These methods belong where the types they act on are. Let's put them in `types.go`. Better still, I'd rename `types.go` to `state.go`

File: records/state/utils.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		ports := yankPorts(t.Resources.Ports)
+		for _, port := range ports {
+			f(port)
+		}
+	}
+}
+
+// ContainerIP extracts a container ip from a Mesos state.json task. If not
+// container ip is provided, an empty string is returned.
+func (t *Task) ContainerIP() string {
```
Comment (tsenart) [inline records/state/utils.go]: I know you just moved these methods, but there's no reason to pass a pointer to a Task when we're not mutating it.

File: records/state/utils_test.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	}
+}
+
+func TestRangePorts(t *testing.T) {
+	p := "[31115-31117]"
+
+	ports := yankPorts(p)
+
+	if len(ports) != 3 {
+		t.Error("not parsing ports")
```
Comment (tsenart) [inline records/state/utils_test.go]: You should `Fatal` here, otherwise the following code might panic.

File: resolver/resolver_test.go [hunk 9]
```diff
@@ -419,13 +420,12 @@ func fakeDNS(t *testing.T) *Resolver {
...
 	}
 
-	var sj records.StateJSON
+	var sj state.State
 	err = json.Unmarshal(b, &sj)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	res.config.Masters = []string{"144.76.157.37:5050"}
```
Comment (tsenart) [inline resolver/resolver_test.go]: What happened here?

File: records/state/utils.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+		for t := lo; t <= hi; t++ {
+			yports = append(yports, strconv.Itoa(t))
+		}
+	}
+	return yports
+}
+
+// ForEachPort calls the given function f for each defined port in the given task
+func (t *Task) ForEachPort(f func(port string)) {
```
Comment (tsenart) [inline records/state/utils.go]: As previously mentioned, let's just have a `func (t Task) Ports() []string`.

File: records/generator.go (hunk 4 above)
Comment (tsenart) [inline records/generator.go]: Something like `func CanonicalName(taskName, taskID, slaveID, frameworkName, domain string) string`

File: records/generator.go (hunk 2 above)
Comment (sttts) [inline records/generator.go]: you are right.

File: records/generator.go (hunk 3 above)
Comment (sttts) [inline records/generator.go]: looks good

File: records/generator.go (hunk 4 above)
Comment (sttts) [inline records/generator.go]: This will get replaced by patterns. Just leave it like it is here.

File: records/generator_test.go (hunk 5 above)
Comment (sttts) [inline records/generator_test.go]: right

File: records/state/utils.go (hunk 6 above)
Comment (sttts) [inline records/state/utils.go]: yes, please. In fact it was the trash bin while refactoring initially ;-)

File: records/state/utils.go (hunk 7 above)
Comment (sttts) [inline records/state/utils.go]: What do you propose? t Task? a task is a pretty big struct. Is go clever enough not to copy everything?

File: records/state/utils_test.go (hunk 8 above)
Comment (sttts) [inline records/state/utils_test.go]: in fact it's the old code, only moved. But please change it.

File: resolver/resolver_test.go (hunk 9 above)
Comment (sttts) [inline resolver/resolver_test.go]: It's in the config already above.

File: records/state/utils.go (hunk 10 above)
Comment (sttts) [inline records/state/utils.go]: same question as above.

**Comment by tsenart** [comment]
LGTM

**Pull request merged**